integration-healthcheck/src/main/java/org/symphonyoss/integration/healthcheck/AsyncCompositeHealthIndicator.java

```diff
@@ -0,0 +1,185 @@
+
+  /**
+   * Details for error message
+   */
+  private static final String ERROR_KEY = "error";
+
+  /**
+   * Thread pool size
+   */
+  private static final Integer MAX_POOL_SIZE = 10;
+
+  /**
+   * Timeout to verify if the execution was done
+   */
+  private static final Long EXECUTION_TIMEOUT = 1000L;
+
+  /**
+   * Number of times to check if the execution was done
+   */
+  private static final Integer SHUTDOWN_CHECK_TIMES = 10;
+
+  /**
+   * Registered indicators
+   */
+  private final Map<String, HealthIndicator> indicators;
+
+  /**
+   * Health aggregator
+   */
+  private final HealthAggregator healthAggregator;
+
+  public AsyncCompositeHealthIndicator(HealthAggregator aggregator) {
+    this.healthAggregator = aggregator;
+    this.indicators = new HashMap<>();
+  }
+
+  /**
+   * Register new indicator
+   * @param name Health indicator name
+   * @param indicator Health indicator object
+   */
+  public void addHealthIndicator(String name, HealthIndicator indicator) {
+    this.indicators.put(name, indicator);
+  }
+
+  @Override
+  public Health health() {
+    try {
+      Map<String, Future<Health>> result = asyncExecution();
+      Map<String, Health> healths = extractResult(result);
+      return this.healthAggregator.aggregate(healths);
+    } catch (InterruptedException e) {
+      String message = "Request thread was interrupted";
+      LOG.error(message, e);
+      return Health.down().withDetail(ERROR_KEY, message).build();
+    }
+  }
+
+  /**
+   * Executes registered indicators using asynchronous calls.
+   * @return
+   * @throws InterruptedException
+   */
+  private Map<String, Future<Health>> asyncExecution() throws InterruptedException {
+    Map<String, Future<Health>> result = new LinkedHashMap<>();
+
+    int indicatorSize = Math.max(1, indicators.size());
+    int poolSize = Math.min(MAX_POOL_SIZE, indicatorSize);
+    ExecutorService service = Executors.newFixedThreadPool(poolSize);
+
+    for (Map.Entry<String, HealthIndicator> entry : indicators.entrySet()) {
+      final HealthIndicator indicator = entry.getValue();
+
+      Future<Health> execution = service.submit(new Callable<Health>() {
+        @Override
+        public Health call() throws Exception {
+          return indicator.health();
+        }
+      });
+
+      result.put(entry.getKey(), execution);
+    }
+
+    shutdownService(service);
+
+    return result;
+  }
+
+  /**
+   * Shutdown the executor service
+   * @param service Executor Service object
+   * @throws InterruptedException Thread execution interrupted
+   */
+  private void shutdownService(ExecutorService service) throws InterruptedException {
+    service.shutdown();
+
+    for (int execution = 0; execution < SHUTDOWN_CHECK_TIMES; execution++) {
+      boolean terminated = service.awaitTermination(EXECUTION_TIMEOUT, TimeUnit.MILLISECONDS);
+
+      if (terminated) {
+        break;
+      }
```

Comment:
Somewhere, you probably need to force termination of threads that don't obey the shutdown() request.  Check isTerminated() and if not true, call shutdownNow().  Or move terminated flag declaration outside the loop and check value after loop is complete.  If false, call shutdownNow().